### PR TITLE
codeintel: RFC 235: Add migration infrastructure to codeintel database

### PR DESCRIPTION
**Pull request by efritz**
A step towards [RFC 235: Move LSIF data to Postgres](https://docs.google.com/document/u/1/d/1Y9p29hK8xrPUTvBdWqP9uAHDg3JLV-HCaxsbCo9-YHQ). Closes https://github.com/sourcegraph/sourcegraph/issues/13885.

Adds initial schema for the new code intel database. Every test location that loads the frontend migrations will also load the code intel migrations.

### CODE REVIEW

File: internal/db/dbstore_db_test.go [hunk 1]
```diff
@@ -17,16 +17,34 @@ func TestMigrations(t *testing.T) {
...
 	}
-	if err := dbutil.DoMigrate(m); err != nil {
-		t.Errorf("error running up migrations: %s", err)
+
+	for _, databaseName := range dbutil.DatabaseNames {
+		t.Run(databaseName, func(t *testing.T) {
+			// Dropping squashed schemas all the way down just drop the entire
+			// database, so when we're testing with a "combined" database that
+			// contains multipleschemas we need to be a bit more careful about
+			// the state after down migrations.
```
Comment (unknwon) [inline internal/db/dbstore_db_test.go:33]: It is not very clear to me the intent. Is our goal to _drop the entire database_ or _drop until the squashed schemas_?

File: internal/db/dbutil/dbutil.go [hunk 2]
```diff
@@ -112,18 +113,52 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+	"codeintel": {
+		MigrationsTable: "codeintel_schema_migrations",
+		Resource:        bindata.Resource(codeintelMigrations.AssetNames(), codeintelMigrations.Asset),
+	},
 }
 
-func NewMigrate(db *sql.DB, dataSource string) (*migrate.Migrate, error) {
-	var cfg postgres.Config
-	driver, err := postgres.WithInstance(db, &cfg)
+var DatabaseNames = func() []string {
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go:130]: nit: Adds a docstring.

File: internal/db/dbutil/dbutil.go [hunk 3]
```diff
@@ -112,18 +113,52 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+	var names []string
+	for databaseName := range databases {
+		names = append(names, databaseName)
+	}
+
+	return names
+}()
+
+var MigrationTables = func() []string {
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go:139]: nit: Adds a docstring.

File: internal/db/dbutil/dbutil.go [hunk 4]
```diff
@@ -112,18 +113,52 @@ func NewDB(dsn, app string) (*sql.DB, error) {
...
+	var migrationTables []string
+	for _, db := range databases {
+		migrationTables = append(migrationTables, db.MigrationsTable)
+	}
+
+	return migrationTables
+}()
+
+func NewMigrate(db *sql.DB, databaseName string) (*migrate.Migrate, error) {
```
Comment (unknwon) [inline internal/db/dbutil/dbutil.go:148]: nit: Adds a docstring.

File: migrations/README.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
+# Postgres Migrations
+
+The children of this directory contain migrations for each Postgres database instance:
```
Comment (unknwon) [inline migrations/README.md:3]: We in fact only have one Postgres database instance, right? Seems not accurate here. :thinking: 

Maybe 
```suggestion
The children of this directory contain migrations for each application that connects to Postgres database:
```

**Review by unknwon — APPROVED** [review]
LGTM.

Curious to know why can't we just rely on frontend to handle all the migrations though. Mind sharing encountered problems with that approach?

File: migrations/README.md (hunk 5 above)
Comment (efritz) [reply migrations/README.md:3]: No, we will have two going forward: https://github.com/sourcegraph/sourcegraph/pull/13904

**Comment by efritz** [comment]
@unknwon 

> Curious to know why can't we just rely on frontend to handle all the migrations though. Mind sharing encountered problems with that approach?

It's a second database instance entirely. See [RFC 235](https://docs.google.com/document/u/1/d/1Y9p29hK8xrPUTvBdWqP9uAHDg3JLV-HCaxsbCo9-YHQ) for details.

File: migrations/README.md (hunk 5 above)
Comment (unknwon) [reply migrations/README.md:3]: Gotcha, thanks!

File: internal/db/dbstore_db_test.go (hunk 1 above)
Comment (efritz) [reply internal/db/dbstore_db_test.go]: I updated these docs. Does it make more sense now?

**Pull request merged**